# Class 11: Derivatives 

Ch. 5, sections 10 - 12: Derivatives, Second derivatives, & Interpolation    

- Complete the activities as instructed by the professor

#### Import any packages we need below (update as we go):

In [1]:
import numpy as np
import math

### Forward, Backward, and Central difference 

Evaluate the derivative of $f(x) = e^{\sin(2x)}$ at $x = 0.5$

If you took the derivative by hand, what would it be? 

$$\frac{df}{dx}=2\text{e}^{\sin{2x}}\cos{2x}$$
$$\left.\frac{df}{dx}\right|_{0.5}\hspace{-2mm}=2\text{e}^{\sin{1}}\cos{1}\approx 2.506761534986894$$

In [9]:
def e_sin(x):
    return np.exp(np.sin(2*x))
def e_sin_derv(x):
    return np.exp(np.sin(2*x))*2*np.cos(2*x)
true_val = e_sin_derv(0.5)
print(true_value)

2.506761534986894


Now apply forward/backward difference 

In [49]:
def fwd_diff(x,h,f):
    return (f(x+h)-f(x))/h

x_fwd_diff = 0.5
h_fwd_diff = np.array([10**(-i) for i in range(1,12)])

fwd_appr_val = fwd_diff(x_fwd_diff,h_fwd_diff,e_sin)

print("the true value is ", true_val)
print(" h      forward diff        error")
rowf = "{0:1.0e} {1:1.16f} {2:1.16f}"

for i in range(len(fwd_appr_val)):
    fwd_err = abs(fwd_appr_val[i]-true_val)
    print(rowf.format(h_fwd_diff[i], fwd_appr_val[i], fwd_err))


the true value is  2.506761534986894
 h      forward diff        error
1e-01 2.1990570766492690 0.3077044583376249
1e-02 2.4807256192968197 0.0260359156900742
1e-03 2.5042064928371133 0.0025550421497806
1e-04 2.5065065168927703 0.0002550180941236
1e-05 2.5067360380326420 0.0000254969542519
1e-06 2.5067589857208361 0.0000025492660578
1e-07 2.5067612785534266 0.0000002564334673
1e-08 2.5067615538887367 0.0000000189018428
1e-09 2.5067619091601046 0.0000003741732106
1e-10 2.5067636855169440 0.0000021505300500
1e-11 2.5067947717616335 0.0000332367747395


backward diff:

In [48]:
def bwd_diff(x,h,f):
    return (-f(x-h)+f(x))/h

x_bwd_diff = 0.5
h_bwd_diff = np.array([10**(-i) for i in range(1,12)])

bwd_appr_val = bwd_diff(x_bwd_diff,h_bwd_diff,e_sin)

print("the true value is ", true_val)
print(" h      backward diff        error")
rowf = "{0:1.0e} {1:1.16f} {2:1.16f}"

for i in range(len(bwd_appr_val)):
    bwd_err = abs(bwd_appr_val[i]-true_val)
    print(rowf.format(h_bwd_diff[i], bwd_appr_val[i], bwd_err))


the true value is  2.506761534986894
 h      backward diff        error
1e-01 2.7076817455157842 0.2009202105288903
1e-02 2.5317171676925376 0.0249556327056437
1e-03 2.5093057730525281 0.0025442380656342
1e-04 2.5070164450369248 0.0002549100500309
1e-05 2.5067870308426166 0.0000254958557226
1e-06 2.5067640843090544 0.0000025493221605
1e-07 2.5067617892560179 0.0000002542691240
1e-08 2.5067615538887367 0.0000000189018428
1e-09 2.5067614650708947 0.0000000699159992
1e-10 2.5067592446248455 0.0000022903620485
1e-11 2.5067503628406484 0.0000111721462455


Now apply central difference 

In [50]:
def cent_diff(x,h,f):
    return (f(x+h)-f(x-h))/(2*h)

x_cent_diff = 0.5
h_cent_diff = np.array([10**(-i) for i in range(1,12)])

cd_appr_val = cent_diff(x_cent_diff,h_cent_diff,e_sin)

print("the true value is ", true_val)
print(" h      central diff        error")
rowf = "{0:1.0e} {1:1.16f} {2:1.16f}"

for i in range(len(cd_appr_val)):
    cd_err = abs(cd_appr_val[i]-true_val)
    print(rowf.format(h_cent_diff[i], cd_appr_val[i], cd_err))

the true value is  2.506761534986894
 h      central diff        error
1e-01 2.4533694110825266 0.0533921239043673
1e-02 2.5062213934946787 0.0005401414922153
1e-03 2.5067561329448207 0.0000054020420732
1e-04 2.5067614809648475 0.0000000540220464
1e-05 2.5067615344376293 0.0000000005492646
1e-06 2.5067615350149453 0.0000000000280513
1e-07 2.5067615339047222 0.0000000010821717
1e-08 2.5067615538887367 0.0000000189018428
1e-09 2.5067616871154996 0.0000001521286057
1e-10 2.5067614650708947 0.0000000699159992
1e-11 2.5067725673011410 0.0000110323142470


How do they compare? 

### Second derivatives

Let's take the second derivative use central difference of the same function $f(x) = e^{\sin(2x)}$ at $x = 0.5$ & compare with the true value 

$$ \frac{d^2}{dx^2}e^{\sin(2x)} = 2 e^{\sin(2x)}(-2\sin(2x)+\cos(4x)+1)  $$

In [58]:
def sec_diff(x,h,f):
    return (f(x+h)+f(x-h)-2*f(x))/(h**2)
def e_sin_derv_2(x):
    return 2*np.exp(np.sin(2*x))*(-2*np.sin(2*x)+np.cos(4*x)+1)

x_sec_diff = 0.5
h_sec_diff = np.array([10**(-i) for i in range(1,12)])

true_val_doub = e_sin_derv_2(0.5)

appr_val_doub = sec_diff(x_sec_diff,h_sec_diff,e_sin)

print("the true value is ", true_val_doub)
print(" h      second diff        error")
rowf = "{0:1.0e} {1:1.16f} {2:1.16f}"

for i in range(len(appr_val_doub)):
    sec_err = abs(appr_val_doub[i]-true_val_doub)
    print(rowf.format(h_sec_diff[i], appr_val_doub[i], sec_err))

the true value is  -5.099281481682785
 h      second diff        error
1e-01 -5.0862466886651516 0.0130347930176331
1e-02 -5.0991548395717956 0.0001266421109891
1e-03 -5.0992802158589257 0.0000012658238591
1e-04 -5.0992814415451448 0.0000000401376399
1e-05 -5.0992809974559341 0.0000004842268506
1e-06 -5.0981441290787188 0.0011373526040659
1e-07 -5.1514348342607272 0.0521533525779425
1e-08 0.0000000000000000 5.0992814816827847
1e-09 0.0000000000000000 5.0992814816827847
1e-10 0.0000000000000000 5.0992814816827847
1e-11 0.0000000000000000 5.0992814816827847


### Automatic Differentiation - book example

Store value/derivative pairs as two-element lists in Python of the form [value, derivative]

Define some functions to perform basic mathematics operations on these inputted lists. 

Apply this to overall figure out the value of the function $f(x) = 3x^2 +e^x$ and its derivative at $t = 1/2$

To do this, perform the following sequence of operations: 

$$ u_1 = x^2$$
$$ u_2 = 3u_1$$
$$ u_3 = e^x$$

and finally 

$$ f(x) = u_2 + u_3$$

Each each mathematical operation will be performed using the generalized operations listed above and the end result will be a pair $[f,f']$, representing the value of $f$ and its derivative.

Final bit of info: to calculate the derivative $f'(t)$ at any value of $t$, need an input list to evaluate our function $f(𝑥)$. Use $[t,1]$ since for whatever value of $t$ we are interested in, by definition, this is the correct value/derivative pair for the independent variable $t$, because $dt/dt = 1$ always. So if we use this pair as the argument of our function, we will get out a pair representing the value of the function $f(t)$ and its derivative $f'(t)$, at the chosen value of $t$.


In [67]:
def power(x,p):
    val_x,dx =  x # x = [x,dx]
    return [val_x**p, p*val_x**(p-1) * dx]
    
def multiplication(x,c):
    val_x,dx = x # x = [x,dx]
    return [c*val_x, c*dx]

def exponential(x):
    val_x,dx =  x # x = [x,dx]
    return [math.exp(val_x), math.exp(val_x) * dx]

def addition(x,y):
    val_x,dx = x
    val_y,dy = y
    return [val_x + val_y, dx + dy]

def Auto_Differentiation(x):
    u1 = power(x,2)
    u2 = multiplication(u1,3)
    u3 = exponential(x)
    u4 = addition(u2,u3)
    return u4

print(Auto_Differentiation([0.5,1]))

[2.398721270700128, 4.648721270700128]


The actual value of the function and the derivative at $f(t)$ at $t = 1/2$ is 
$$f (t) = 3t^2 + e^t = \frac{3}{4} + e^{1/2} = 2.3987 ...$$
$$ f'(t) = 6t + e^t = 3 + e^{1/2} = 4.6487 ...$$

Look how exact we are!